File: src/linux/src/StepperPRU.cpp
/* StepperPRU.cpp — Spindle stepper API delegating to PRU0 via IpcChannel.
 *
 * Klipper-style implementation: step moves pre-computed on host, pushed to
 * the PRU move ring.  No Hz/accel commands sent to PRU; the IEP counter
 * handles precise edge timing.
 *
 * Compensation (spindle-lateral ratio) remains host-side:
 *   WinderApp calls setSpeedHz(compensatedHz) every control loop tick.
 *   tick() detects the changed target and transitions via accel segment.
 */

#include "StepperPRU.h"
#include <cmath>

static inline uint32_t clamp_hz(uint32_t v, uint32_t lo, uint32_t hi) {
    if (v < lo) return lo;
    if (v > hi) return hi;
    return v;
}

StepperPRU::StepperPRU(IpcChannel &channel) : _channel(channel) {}

void StepperPRU::begin(uint32_t accelHzPerMs) {
    if (!_channel.isOpen()) return;
    _accelHzPerS = accelHzPerMs * 1000u;
    _channel.sendCommand(CMD_ENABLE, AXIS_SPINDLE, 0u);   /* driver disabled */
}

void StepperPRU::setSpeedHz(uint32_t hz) {
    uint32_t clamped = clamp_hz(hz, SPEED_HZ_MIN, SPEED_HZ_MAX);
    if (clamped == _speedHz) return;
    _speedHz     = clamped;
    _speedChanged = true;
}

void StepperPRU::start(bool forward) {
    if (!_channel.isOpen()) return;
    _forward   = forward;
    _streaming = true;
    _channel.sendCommand(CMD_ENABLE, AXIS_SPINDLE, 1u);

    /* Push accel segment from SPEED_HZ_MIN to _speedHz */
    uint32_t startHz = SPEED_HZ_MIN;
    if (_speedHz > startHz) {
        MoveQueue::pushAccelSegment(_channel, AXIS_SPINDLE,
                                    startHz, _speedHz, forward, _accelHzPerS);
    }
    /* Seed the ring with initial constant-speed block */
    MoveQueue::pushConstantMs(_channel, AXIS_SPINDLE,
                               _speedHz, forward, REFILL_MS * KEEP_SLOTS);
    _pushedHz     = _speedHz;
    _speedChanged = false;
}

void StepperPRU::stop() {
    if (!_channel.isOpen()) return;
    _streaming = false;
    MoveQueue::pushDecelToStop(_channel, AXIS_SPINDLE,
                                _pushedHz, _forward, _accelHzPerS);
    _pushedHz = 0u;
}

void StepperPRU::forceStop() {
    if (!_channel.isOpen()) return;
    _streaming = false;
    _channel.sendCommand(CMD_QUEUE_FLUSH, AXIS_SPINDLE);
    _channel.sendCommand(CMD_ENABLE,      AXIS_SPINDLE, 0u);
    _pushedHz = 0u;
}

void StepperPRU::disableDriver() {
    if (!_channel.isOpen()) return;
    _channel.sendCommand(CMD_ENABLE, AXIS_SPINDLE, 0u);
}

bool StepperPRU::isRunning() const {
    _refreshTelem();
    return (_telem.state & STATE_RUNNING) != 0u;
}

long StepperPRU::getTurns() const {
    _refreshTelem();
    return (long)(_telem.step_count / STEPS_PER_REV);
}

void StepperPRU::resetTurns() {
    if (!_channel.isOpen()) return;
    _channel.sendCommand(CMD_RESET_POSITION, AXIS_SPINDLE);
}

float StepperPRU::getRPM() const {
    _refreshTelem();
    uint32_t iv = _telem.current_interval;
    if (iv == 0u) return 0.0f;
    /* step_hz = PRU_CLOCK/(2*interval); rpm = hz*60/SPR */
    float hz = (float)PRU_CLOCK_HZ / (2.0f * (float)iv);
    return hz * 60.0f / (float)STEPS_PER_REV;
}

void StepperPRU::tick() {
    if (!_streaming || !_channel.isOpen()) return;
    _refreshTelem();

    /* Speed transition: flush ring and ramp to new target */
    if (_speedChanged) {
        _channel.sendCommand(CMD_QUEUE_FLUSH, AXIS_SPINDLE);
        uint32_t from = (_pushedHz > 0u) ? _pushedHz : SPEED_HZ_MIN;
        if (_speedHz != from) {
            MoveQueue::pushAccelSegment(_channel, AXIS_SPINDLE,
                                        from, _speedHz, _forward, _accelHzPerS);
        }
        /* Seed constant-speed block after ramp to prevent underrun */
        MoveQueue::pushConstantMs(_channel, AXIS_SPINDLE,
                                   _speedHz, _forward, REFILL_MS * KEEP_SLOTS);
        _pushedHz     = _speedHz;
        _speedChanged = false;
    }

    /* Refill constant-speed block when ring is low */
    uint32_t slots = _channel.moveQueueFreeSlots(AXIS_SPINDLE);
    if (slots >= IPC_SP_MOVE_SLOTS - KEEP_SLOTS) {
        MoveQueue::pushConstantMs(_channel, AXIS_SPINDLE,
                                   _pushedHz, _forward, REFILL_MS);
    }
}

void StepperPRU::_refreshTelem() const {
    _channel.readSpindleTelem(_telem);
}

uint32_t StepperPRU::calculateCompensatedSpindleHz(uint32_t nominalSpindleHz,
                                                    uint32_t currentLatHz,
                                                    uint32_t nominalLatHz) {
    if (nominalLatHz == 0u || nominalSpindleHz == 0u || currentLatHz < 100u)
        return nominalSpindleHz;
    uint32_t comp = (uint32_t)(((uint64_t)nominalSpindleHz * currentLatHz)
                               / nominalLatHz);
    return clamp_hz(comp, StepperPRU::SPEED_HZ_MIN, StepperPRU::SPEED_HZ_MAX);
}

File: src/python/stepper_pru.py
"""StepperPRU - spindle axis wrapper around IpcChannel.

Mirrors StepperPRU.cpp / StepperPRU.h exactly.
Klipper-style: all steps are pushed as (interval, count, add) move triples.
No Hz/accel/start/stop commands are sent to the PRU.
"""
from ipc import (
    IpcChannel,
    CMD_ENABLE, CMD_EMERGENCY_STOP, CMD_RESET_POSITION, CMD_QUEUE_FLUSH,
    AXIS_SPINDLE, AXIS_ALL,
    STATE_RUNNING, STATE_ENABLED,
)
from config import STEPS_PER_REV, SPEED_HZ_MIN, SPEED_HZ_MAX

PRU_CLOCK_HZ = 200_000_000
MIN_INTERVAL  = 625          # = PRU_CLOCK_HZ / (2 * 160000)
_LAT_MIN_HZ_THRESHOLD = 100
_REFILL_MS = 10
_KEEP_SLOTS = 4


def _clamp_hz(v: int) -> int:
    return max(SPEED_HZ_MIN, min(SPEED_HZ_MAX, v))


def _hz_to_interval(hz: int) -> int:
    if hz <= 0:
        return 0
    iv = PRU_CLOCK_HZ // (2 * hz)
    return max(MIN_INTERVAL, iv)


def _push_constant_ms(ch: IpcChannel, hz: int, forward: bool, duration_ms: int) -> bool:
    """Push a constant-speed block for approximately duration_ms."""
    if hz <= 0 or duration_ms <= 0:
        return False
    iv = _hz_to_interval(hz)
    # edges per ms = 2*hz/1000, total edges for duration
    edges = max(1, (2 * hz * duration_ms) // 1000)
    direction = 0 if forward else 1
    return ch.sendMove(AXIS_SPINDLE, iv, edges, 0, direction)


class StepperPRU:
    """Spindle axis control via PRU IPC.  Mirrors StepperPRU.cpp."""

    def __init__(self, channel: IpcChannel):
        self._ch = channel
        self._speed_hz: int = 0
        self._pushed_hz: int = 0
        self._forward: bool = True
        self._streaming: bool = False
        self._speed_changed: bool = False
        self._accel_hz_per_s: int = 150_000
        self._telem: dict = {}

    def begin(self, accel_hz_per_ms: int = 150):
        if not self._ch.isOpen():
            return
        self._accel_hz_per_s = accel_hz_per_ms * 1000
        self._ch.sendCommand(CMD_ENABLE, AXIS_SPINDLE, 0)  # disabled at boot

    #  Telemetry
    def _refresh(self):
        self._telem = self._ch.readSpindleTelem()

    def is_running(self) -> bool:
        self._refresh()
        return bool(self._telem.get('state', 0) & STATE_RUNNING)

    def is_enabled(self) -> bool:
        self._refresh()
        return bool(self._telem.get('state', 0) & STATE_ENABLED)

    def get_turns(self) -> int:
        self._refresh()
        sc = self._telem.get('step_count', 0)
        return sc // STEPS_PER_REV if STEPS_PER_REV else 0

    def get_rpm(self) -> float:
        self._refresh()
        iv = self._telem.get('current_interval', 0)
        if not iv:
            return 0.0
        hz = PRU_CLOCK_HZ / (2 * iv)
        return hz * 60.0 / STEPS_PER_REV if STEPS_PER_REV else 0.0

    #  Control
    def set_speed_hz(self, hz: int):
        clamped = _clamp_hz(int(hz))
        if clamped == self._speed_hz:
            return
        self._speed_hz = clamped
        self._speed_changed = True

    def start(self, forward: bool = True):
        if not self._ch.isOpen():
            return
        self._forward = forward
        self._streaming = True
        self._ch.sendCommand(CMD_ENABLE, AXIS_SPINDLE, 1)
        _push_constant_ms(self._ch, self._speed_hz, forward, _REFILL_MS * _KEEP_SLOTS)
        self._pushed_hz = self._speed_hz
        self._speed_changed = False

    def stop(self):
        if not self._ch.isOpen():
            return
        self._streaming = False
        # Push zero-speed block (decel would need MoveQueue ramp math)
        self._ch.sendCommand(CMD_QUEUE_FLUSH, AXIS_SPINDLE)
        self._pushed_hz = 0

    def force_stop(self):
        if not self._ch.isOpen():
            return
        self._streaming = False
        self._ch.sendCommand(CMD_QUEUE_FLUSH, AXIS_SPINDLE)
        self._ch.sendCommand(CMD_ENABLE, AXIS_SPINDLE, 0)
        self._pushed_hz = 0

    def disable(self):
        if not self._ch.isOpen():
            return
        self._ch.sendCommand(CMD_ENABLE, AXIS_SPINDLE, 0)

    def reset_turns(self):
        if not self._ch.isOpen():
            return
        self._ch.sendCommand(CMD_RESET_POSITION, AXIS_SPINDLE)

    def emergency_stop(self):
        if not self._ch.isOpen():
            return
        self._ch.sendCommand(CMD_EMERGENCY_STOP, AXIS_ALL)

    def tick(self):
        """Refill move ring; handle speed changes.  Call at control cadence."""
        if not self._streaming or not self._ch.isOpen():
            return
        if self._speed_changed:
            self._ch.sendCommand(CMD_QUEUE_FLUSH, AXIS_SPINDLE)
            _push_constant_ms(self._ch, self._speed_hz, self._forward,
                               _REFILL_MS * _KEEP_SLOTS)
            self._pushed_hz = self._speed_hz
            self._speed_changed = False
        else:
            free = self._ch.moveQueueFreeSlots(AXIS_SPINDLE)
            if free >= 4:
                _push_constant_ms(self._ch, self._pushed_hz, self._forward, _REFILL_MS)

    @staticmethod
    def calculate_compensated_spindle_hz(
        nominal_spindle_hz: int,
        current_lat_hz: int,
        nominal_lat_hz: int
    ) -> int:
        """Scale spindle Hz proportionally to actual lateral Hz."""
        if nominal_lat_hz == 0 or nominal_spindle_hz == 0:
            return nominal_spindle_hz
        if current_lat_hz < _LAT_MIN_HZ_THRESHOLD:
            return nominal_spindle_hz
        comp = (nominal_spindle_hz * current_lat_hz) // nominal_lat_hz
        return _clamp_hz(comp)
